### PR TITLE
Issue #3702: Allow single character names in local variables, method and catch-blocks parameters names in accordance with Google Style Guide

### PR DESCRIPTION
**Pull request by MEZk**
#3702

### CODE REVIEW

**Comment by romani** [comment]
diff report on guava is required.

please remove fix for Abbreviation, as 0 value lead to unclear violation message. See #3721 .

**Comment by MEZk** [comment]
@romani 
>diff report on guava is required.

Done. See http://mezk.github.io/i3072-google-config/diff

> please remove fix for Abbreviation, as 0 value lead to unclear violation message.

Done

**Comment by MEZk** [comment]
@romani 
CIs are green.

File: src/it/resources/com/google/checkstyle/test/chapter5naming/rule51identifiernames/InputCatchParameterName.java [hunk 1]
```diff
@@ -21,7 +21,7 @@
         } catch (Exception noWorries) { // ok
         }
         try {
-        } catch (Throwable t) { // warn
+        } catch (Throwable t) {
```
Comment (romani) [inline src/it/resources/com/google/checkstyle/test/chapter5naming/rule51identifiernames/InputCatchParameterName.java]: please put "// ok" instead, not sure why we did this .... but lets keep this.

File: src/main/resources/google_checks.xml [hunk 2]
```diff
@@ -109,26 +109,26 @@
...
-            <property name="format" value="^[a-z]([a-z0-9][a-zA-Z0-9]*)?$"/>
+            <property name="format" value="^[a-z]([a-z0-9]+[A-Z]?)*$"/>
             <property name="accessModifiers" value="protected, package, private"/>
             <message key="name.invalidPattern"
              value="Parameter name ''{0}'' must match pattern ''{1}''."/>
         </module>
         <module name="ParameterName">
             <property name="id" value="ParameterNamePublic"/>
-            <property name="format" value="^[a-z][a-z0-9][a-zA-Z0-9]*$"/>
+            <property name="format" value="^[a-z]([a-z0-9][A-Z]?)+$"/>
```
Comment (romani) [inline src/main/resources/google_checks.xml]: I am not sure why non-public and public regexps are different. Google confirmed two times in issue on google style that checkstyle rule should be relaxed. It is their style - so we follow it. If smb do not like this (I hope most of projects) they will change this in their configuration.

`accessModifiers` should be removed from google style config.

Comment (MEZk) [inline src/main/resources/google_checks.xml]: @romani 
Google style prohibits one-character parameter names for public methods and allows them form non-public methods. That is why we have 2 different regexpses.

File: src/it/resources/com/google/checkstyle/test/chapter5naming/rule51identifiernames/InputCatchParameterName.java (hunk 1 above)
Comment (MEZk) [inline src/it/resources/com/google/checkstyle/test/chapter5naming/rule51identifiernames/InputCatchParameterName.java]: @romani 
Done.

File: src/main/resources/google_checks.xml (hunk 2 above)
Comment (romani) [inline src/main/resources/google_checks.xml]: Please read the issue discussion on Google style guide https://github.com/google/styleguide/issues/214#issuecomment-271221185

Comment (rnveach) [inline src/main/resources/google_checks.xml]: @romani Should we re-open issue and request the documentation be changed?
We have said before we only go by that documentation like when they requested us to validate against another tool.

File: src/it/java/com/google/checkstyle/test/chapter5naming/rule51identifiernames/CatchParameterNameTest.java [hunk 3]
```diff
@@ -22,29 +22,35 @@
...
     public void catchParameterNameTest() throws Exception {
-        final Configuration checkConfig = getCheckConfig("CatchParameterName");
         final String msgKey = "name.invalidPattern";
-        final String format = "^[a-z][a-z0-9][a-zA-Z0-9]*$";
 
         final String[] expected = {
-            "6:28: " + getCheckMessage(checkConfig.getMessages(), msgKey, "e", format),
-            "24:28: " + getCheckMessage(checkConfig.getMessages(), msgKey, "t", format),
             "47:28: " + getCheckMessage(checkConfig.getMessages(), msgKey, "iException", format),
-            "50:28: " + getCheckMessage(checkConfig.getMessages(), msgKey, "x", format),
```
Comment (rnveach) [inline src/it/java/com/google/checkstyle/test/chapter5naming/rule51identifiernames/CatchParameterNameTest.java]: One example of a violation isn't much. Can we change or add more examples?

File: src/main/resources/google_checks.xml (hunk 2 above)
Comment (romani) [inline src/main/resources/google_checks.xml]: Reopen will not work, let's do reopen and PR to propose the fix

Comment (MEZk) [inline src/main/resources/google_checks.xml]: @romani @rnveach 
I opened the issue.
https://github.com/google/styleguide/issues/226
I will propose PR, if they accept the issue.

File: src/it/java/com/google/checkstyle/test/chapter5naming/rule51identifiernames/CatchParameterNameTest.java (hunk 3 above)
Comment (MEZk) [inline src/it/java/com/google/checkstyle/test/chapter5naming/rule51identifiernames/CatchParameterNameTest.java]: @rnveach 
Done.

**Comment by romani** [comment]
`accessModifiers` usage has to be removed.

**Comment by MEZk** [comment]
@romani 
Google has not accepted my changes to Google Java Style Guide yet. If they accept, we will still need accessMModifiers, as I proposed to following changes: https://github.com/google/styleguide/pull/227/files#diff-b6c9191ee5ccb7a5ef170e98e51d94d8R886

It means that formulation will become more strict, but still specific to **public** methods parameter names.

**Comment by MEZk** [comment]
@romani @rnveach 
Are you OK with the changes?

**Comment by rnveach** [comment]
If we merge this we will be out of sync with the google document for this one item as style team has not responded to Issue or PR.

File: src/main/resources/google_checks.xml [hunk 4]
```diff
@@ -108,27 +108,18 @@
...
-            <property name="format" value="^[a-z][a-z0-9][a-zA-Z0-9]*$"/>
+            <property name="format" value="^[a-z]([a-z0-9]+[A-Z]?)*$"/>
             <message key="name.invalidPattern"
              value="Catch parameter name ''{0}'' must match pattern ''{1}''."/>
         </module>
         <module name="LocalVariableName">
             <property name="tokens" value="VARIABLE_DEF"/>
-            <property name="format" value="^[a-z][a-z0-9][a-zA-Z0-9]*$"/>
+            <property name="format" value="^[a-z]([a-z0-9]+[A-Z]?)*$"/>
             <property name="allowOneCharVarInForLoop" value="true"/>
```
Comment (romani) [inline src/main/resources/google_checks.xml]: looks like we do not need this property `allowOneCharVarInForLoop` anymore , please remove

Comment (MEZk) [inline src/main/resources/google_checks.xml]: @romani 
Done

**Comment by MEZk** [comment]
@romani 
> please regenerate reports to make sure that my last request do not have affect validation.

DIff reports:
1) [ParameterNameCheck](http://mezk.github.io/i3702-google-config/diff/ParameterNameCheck)
2) [LocalVariableNameCheck](http://mezk.github.io/i3702-google-config/diff/LocalVariableName)
3) [CatchParameterNameCheck](http://mezk.github.io/i3702-google-config/diff/CatchParameterName)

**Comment by romani** [comment]
@MEZk , reports are ok, but they are show that you change config in the way that is not requested in issue.
You changed a format to address abbreviations, that was not requested in issue.
If you think it also should be done - please create separate issue on this. Do not mix all fixes together!

1) config regexps should be fixed 
2) reports need to be regenerated

**Comment by MEZk** [comment]
@romani 
> 1. config regexps should be fixed

Done. Config was changed to allow only one-characters names. ParameterNameCheck, LocalVariableCheck and CatchParameterCheck should not validate abbreviations in identifiers names. Abbreviations should be validated by special check AbbreviationAsWordInName. I created separate issue to fix this: https://github.com/checkstyle/checkstyle/issues/3812

> 2. reports need to be regenerated
Done. See http://mesk.github.io/i3702-google-config/diff/

**Comment by romani** [comment]
I canceled MacOS build manually (performance reason), it is ok to skip it for this PR.
code changes is ok, config changes is ok.
report is good.

@rnveach , please review one more time and do merge if all is fine to your mind.

**Pull request merged**